test/testGridConnect.cpp

```diff
@@ -0,0 +1,222 @@
+  testGridConnectDecode_ID(":XFFE3FFFFN;", true, 0x1FFFFFFF, true);   // all bits 0 to 28 set
+  testGridConnectDecode_ID(":XFFFFFFFFN;", true, 0x1FFFFFFF, true);   // check it ignores unused bits
+  testGridConnectDecode_ID(":XQ0000000N;", true, 0x0, false);         // non hex char in ID
+  testGridConnectDecode_ID(":X0000000QN;", true, 0x0, false);         // non hex char in ID
+  
+  // test decode RTR - params are inoput msg, expected RTR, expected result from decode
+  testGridConnectDecode_RTR(":S0000N;", false, true);             // standard msg, RTR false
+  testGridConnectDecode_RTR(":S0000R;", true, true);              // standard msg, RTR true
+  testGridConnectDecode_RTR(":S0000Q;", true, false);             // standard msg, invalid char in RTR field
+  testGridConnectDecode_RTR(":X00000000N;", false, true);         // extended msg, RTR false
+  testGridConnectDecode_RTR(":X00000000R;", true, true);          // extended msg, RTR true
+  testGridConnectDecode_RTR(":X00000000Q;", true, false);          // extended msg, invalid char in RTR field
+
+  // test decode data field, standard message
+  testGridConnectDecode_DATA(":S0000N;", 0, true);                          // standard msg, zero data
+  testGridConnectDecode_DATA(":S0000N01;", 1, true);                        // standard msg, 1 data byte
+  testGridConnectDecode_DATA(":S0000N0102030405060708;", 8, true);          // standard msg, 8 data bytes
+  testGridConnectDecode_DATA(":S0000N0;", 1, false);                        // standard msg, wrong number of data chars
+  testGridConnectDecode_DATA(":S0000N010203040506070;", 8, false);          // standard msg, wrong number of data chars
+  testGridConnectDecode_DATA(":S0000N010203040506070809;", 9, false);       // standard msg, too many data bytes
+  testGridConnectDecode_DATA(":S0000NQ102030405060708;", 8, false);          // standard msg, invalid char in data
+  testGridConnectDecode_DATA(":S0000N010203040506070Q;", 8, false);          // standard msg, invalid char in data
+
+  // test decode data field, extended message
+  testGridConnectDecode_DATA(":X00000000N;", 0, true);                      // extended msg, zero data
+  testGridConnectDecode_DATA(":X00000000N01;", 1, true);                    // extended msg, 1 data byte
+  testGridConnectDecode_DATA(":X00000000N0102030405060708;", 8, true);      // extended msg, 8 data bytes
+  testGridConnectDecode_DATA(":X00000000N0;", 1, false);                    // extended msg, wrong number of data chars
+  testGridConnectDecode_DATA(":X00000000N010203040506070;", 8, false);      // extended msg, wrong number of data chars
+  testGridConnectDecode_DATA(":X00000000N010203040506070809;", 9, false);   // extended msg, too many data bytes
+  testGridConnectDecode_DATA(":X00000000NQ102030405060708;", 8, false);     // extended msg, invalid char in data
+  testGridConnectDecode_DATA(":X00000000N010203040506070Q;", 8, false);     // extended msg, invalid char in data
+
+
+}
```

Comment:
Please also add some tests with lower case characters, both hex and the S/X and N/R bits.

Reply:
Done that - now GridConnect Decodes will fail if lower case present (as per GridConnect format)

Reply:
Hmm, I thought they would succeed. but now I see that the toupper() call is in SerialGC instead.